src/util.hpp

```diff
@@ -29,5 +29,42 @@ inline char* to_csv(const std::vector<std::string>& names) {
   return strdup(s_c);
 }
 
+/**
+ * Convert exception-style error handling into our C output parameter
+ * style. F is always a lambda, so this code gets inlined and the function
+ * call overhead is optimized away.
+ */
+template <typename F>
+[[gnu::always_inline]] [[msvc::forceinline]]
+inline auto handle_errors(const char* name, char** error_msg, F f) {
+  try {
+    return f();
+  } catch (const std::exception& e) {
+    if (error_msg) {
+      std::stringstream error;
+      error << name << "() failed with exception: " << e.what() << std::endl;
+      *error_msg = strdup(error.str().c_str());
+    }
+  } catch (...) {
+    if (error_msg) {
+      std::stringstream error;
+      error << name << "() failed with unknown exception" << std::endl;
+      *error_msg = strdup(error.str().c_str());
+    }
+  }
+
+  // Handle the return value in the failure case, generically
+  using Result = std::invoke_result_t<F>;
+  if constexpr (std::is_same_v<Result, int>) {
```

Comment:
Should we be explicitly sizing all of our `int` values in C++11 and beyond?  I'm not sure what the best practice is.

Reply:
Because the goal of these functions is to be used in FFI, restricting to the “old” C types is the best for compatibility. Julia for example doesn’t define FFI types for the C exact sized integer types

Reply:
Got it.  What a mess all this undefined or partially specified behavior is!